fat/fat2/transactionio.go

```
@@ -0,0 +1,67 @@
+package fat2
+
+import (
+	"fmt"
+	"github.com/Factom-Asset-Tokens/factom"
+)
+
+// TransactionIO represents an individual input or output of a fat2 transaction
+type TransactionIO struct {
+	Address *factom.FAAddress `json:"address,omitempty"`
+	Type    *string           `json:"type,omitempty"`
+	Amount  *uint64           `json:"amount,omitempty"`
+}
+
+// Valid checks that this input/output has at least one non-nil field
+// and has a valid asset type (if present). A nil value is returned if
+// all checks pass. Note: this function evaluates the input/output
+// in a vacuum, any contextual validation happens on a Transaction struct.
+func (txio TransactionIO) Valid() error {
+	hasAddress := txio.Address != nil
+	hasType := txio.Type != nil
+	hasAmount := txio.Amount != nil
+	// At least one of the fields must be non-nil
+	if (hasAddress || hasType || hasAmount) == false {
```

While this is highly readable, it deprives the compiler from failing early when doing these comparisons. It can be made just as readable with appropriate line spacing. Please use this instead:

```
	if txio.Address == nil &&
		txio.Amount == nil &&
		txio.Type == nil {
		return fmt.Errorf("at least one non-nil field required")
	}
```

---

Definitely agree, fixed now